---
title: "Overlapping genes"
author: Gibran Hemani
date: "2023-02-18"
categories: [genomics]
---

## Background

For a list of positions, identify all genes that overlap +/- 200kb.

Use Ensembl database of gene positions 

```{r}
library(EnsDb.Hsapiens.v75) # For convenience hg19
library(ieugwasr)
library(dplyr)
genelist <- genes(EnsDb.Hsapiens.v75) %>%
    subset(gene_biotype=="protein_coding")
genelist
```

SNP list

```{r}
snps_cluster1 <- c("rs1097327","rs2186120","rs2166172","rs75641275","rs12037698","rs1446585","rs16846140","rs13062093","rs2051559","rs6861649","rs2281819","rs12662900","rs9388681","rs17132130","rs215634","rs79682948","rs2192649","rs13294945","rs7357754","rs4749937","rs1465900","rs1799992","rs55938344","rs7987928","rs7331420","rs9522279","rs55689274","rs4777541","rs7189149","rs11079849","rs113230003","rs150998792")
```

Get positions

```{r}
bmi <- associations(snps_cluster1, "ukb-a-248")
target <- GRanges(bmi$chr, IRanges(bmi$position - 200000, bmi$position + 200000))
target
```

Find overlaps with genelist

```{r}
overlaps <- findOverlaps(target, genelist)
overlaps
```

Overall about 116 overlapping protein coding genes found for 30 SNPs.

---

Try with all ensembl annotations, not just protein coding

```{r}
genelist_all <- genes(EnsDb.Hsapiens.v75)
genelist_all
overlaps1 <- findOverlaps(target, genelist_all)
overlaps1
```

Still only 277

---

Try with HumanHT-12 v3.0 probes. Download manifest file to get probe coordinates:

```{r}
temp <- tempfile()
download.file("https://emea.support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/humanht-12/v3/humanht-12_v3_0_r3_11283641_a_txt.zip", temp)
library(data.table)
a <- fread(paste0("unzip -p ", temp), skip=8)
str(a)
a$bp <- sapply(strsplit(a$Probe_Coordinates, "-"), function(o) as.numeric(o[1]))
str(a)
a <- subset(a, !is.na(bp))
dim(a)
b <- GRanges(a$Chromosome, IRanges(a$bp, a$bp), Transcript=a$Transcript)
overlaps2 <- findOverlaps(target, b)
overlaps2
```

Still only 165

---

Looks like the overlaps are actually only for eQTL SNP positions, rather than probe or annotation locations. So get MuTHER eQTL results, and lookup how many eQTLs overlap BMI SNPs +/-200kb

```{r}
temp <- tempfile()
download.file("http://www.muther.ac.uk/DataForWebsite/MuTHER_top_cis_eQTL_per_probe_Fat.txt.gz", temp)
b <- fread(paste0("gunzip -c ", temp))
b
bg <- GRanges(b$CHR, IRanges(b$SNP_Coor, b$SNP_Coor), Gene=b$Gene)
bg
overlaps3 <- findOverlaps(target, bg)
length(unique(b$Gene[overlaps3@to]))
```

Still not much. But I think in the script it suggests that all cis-SNPs were included without filtering for p-value, so actually it's looking for any overlap with any genes +/- 1Mb?

```{r}
bg <- GRanges(b$CHR, IRanges(b$TSS-1000000, b$TSS+1000000), Gene=b$Gene)
bg
overlaps3 <- findOverlaps(target, bg)
length(unique(b$Gene[overlaps3@to]))
```

Now it's quite a it higher.

```{r}
sessionInfo()
```
