---
title: "Spurious vQTL simulation"
author: Gibran Hemani
date: "2023-02-28"
categories: ['genetics', 'interactions', 'statistics']
---

## Background

How sensitive is DRM to incomplete LD in more realistic scenarios? Take a region of 1000 genomes data and choose a SNP at random to have an additive effect of some determined magnitude. We'd now determine how many SNPs in the region (i.e. with incomplete LD with the causal SNP) have evidence of vQTL using DRM or SVLM. How does genetic effect size and sample size relate to minimum p-value and number of positions with FDR < 0.05.

```{r}
#| message: false
#| warning: false

library(glue)
library(data.table)
library(dplyr)
library(ggplot2)
```

```{r}
test_drm <- function(g, y)
{
  y.i <- tapply(y, g, median, na.rm=T)
  z.ij <- abs(y - y.i[g+1])
  summary(lm(z.ij ~ g))$coef %>%
    as_tibble() %>%
    slice(2) %>%
    mutate(method="drm")
}

test_svlm <- function(g, y)
{
  yres <- residuals(lm(y ~ g))^2
  summary(lm(yres ~ g))$coefficients %>%
    as_tibble() %>%
    slice(n=2) %>%
    mutate(method="svlm")
}
```

Get some plink data - e.g. just 1-2k SNPs from 1000 genomes European samples

```{bash}
/Users/gh13047/Downloads/plink_mac_20230116/plink \
    --bfile /Users/gh13047/repo/opengwas-api-internal/opengwas-api/app/ld_files/EUR \
    --chr 22 \
    --recode A \
    --out eur22 \
    --from-bp 0 --to-bp 17000000
```

Read in genotype data

```{r}
geno <- fread("eur22.raw")
fam <- geno[,1:6]
geno <- as.matrix(geno[,-c(1:6)])
geno[1:10,1:10]
dim(geno)

cormat <- cor(geno)
dim(cormat)
```

```{r}
fast_assoc <- function(y, x)
{
    index <- is.finite(y) & is.finite(x)
    n <- sum(index)
    y <- y[index]
    x <- x[index]
    vx <- var(x)
    bhat <- stats::cov(y, x)/vx
    ahat <- mean(y) - bhat * mean(x)
    rsq <- (bhat * vx)^2/(vx * var(y))
    fval <- rsq * (n - 2)/(1 - rsq)
    tval <- sqrt(fval)
    se <- abs(bhat/tval)
    p <- stats::pf(fval, 1, n - 2, lower.tail = FALSE)
    return(list(ahat = ahat, bhat = bhat, se = se, fval = fval, pval = p, n = n))
}

test_drm <- function(g, y)
{
  y.i <- tapply(y, g, median, na.rm=T)
  z.ij <- abs(y - y.i[g+1])
  fast_assoc(z.ij, g) %>% as_tibble()
}

sim_full <- function(rsq, geno, n, cormat)
{
  i <- sample(1:ncol(geno), 1)
  y <- as.numeric(scale(geno[,i])) * sqrt(rsq) + rnorm(nrow(geno), 0, sqrt(1-rsq))
  res <- lapply(1:ncol(geno), function(i)
  {
      test_drm(geno[1:n,i], y[1:n])
  }) %>% bind_rows() %>% mutate(snp=1:n(), ldrsq=cormat[i,]^2)
  return(res)
  res %>% 
    mutate(fdr = p.adjust(pval, "fdr")) %>%
    summarise(
        vqtl = which.min(pval),
        minp = min(pval, na.rm=T),
        nfdr = sum(fdr < 0.05, na.rm=T),
        vqtl_ldrsq = ldrsq[vqtl]
        ) %>%
    mutate(rsq = rsq, n = n, qtl=i, af=sum(geno[,i])/(2*nrow(geno))) %>%
    ungroup()
}
set.seed(1234)
o <- sim_full(0.5, geno, 500, cormat)
o %>% ggplot(., aes(x=ldrsq, y = -log10(pval))) +
geom_point() +
geom_smooth()
```


## Determining spurious effects

Test DRM on each SNP in region for varying sample sizes and additive variance explained in the region

Simulation here: https://github.com/explodecomputer/mz-gwas/blob/main/scripts/drm_sims.r

```{r}
load(url("https://github.com/explodecomputer/mz-gwas/raw/main/scripts/drm_sims.rdata"))

ggplot(res %>% filter(minp < 5e-8), aes(x=as.factor(rsq), y=nfdr)) +
geom_boxplot(aes(fill=as.factor(n)))
```


```{r}
ggplot(res %>% filter(minp < 5e-8), aes(x=as.factor(rsq), y=-log10(minp))) +
geom_boxplot(aes(fill=as.factor(n)))
```


```{r}
ggplot(res %>% filter(minp < 5e-10), aes(x=af, y=vqtl_ldrsq)) +
geom_point()
```





## Summary

- SVLM and DGLM suffer equally from the issue
- The problem is likely a function of F statistic and LD - i.e. as sample sizes get larger, smaller effects additive become more relevant
- With n=500 additive effect of rsq = 0.1 is enough to generate a spurious significant vQTL


```{r}
unlink("eur22*")
```

---

```{r}
sessionInfo()
```
